Tactics/Sym.lean

```diff
@@ -12,93 +12,211 @@ import Tactics.SymContext
-
-/-- A wrapper around `evalTactic` that traces the passed tactic script and
-then executes those tactics -/
-private def evalTacticAndTrace (tactic : TSyntax `tactic) : TacticM Unit := do
-  trace[Sym] "running:\n{tactic}"
-  evalTactic tactic
-
-/-- `init_next_step h_run` splits the hypothesis
+open Lean Meta
+open Lean.Elab.Tactic
 
-`h_run: s_final = run n s`
+/-- A wrapper around `evalTactic` that traces the passed tactic script,
+executes those tactics, and then traces the new goal state -/
+private def evalTacticAndTrace (tactic : TSyntax `tactic) : TacticM Unit :=
+  withTraceNode `Tactic.sym (fun _ => pure m!"running: {tactic}") <| do
+    evalTactic tactic
+    trace[Tactic.sym] "new goal state:\n{← getGoals}"
 
-introduces a new state variable, say `s_next` and two new hypotheses:
-`h_step: s_next = stepi s`
-`h_run': s_final = run (n-1) s_next`
-
-The new state variable and the hypotheses are not named yet.
--/
-macro "init_next_step" h_run:ident h_step:ident sn:ident : tactic =>
+/-- `init_next_step h_run stepi_eq sn` splits the hypothesis
+  `h_run: s_final = run (n+1) s`
+by adding a new state variable, `sn`, and two new hypotheses:
+  `stepi_eq : stepi s = s_next`
+  `h_run'   : s_final = run n s_next`
+to the local context of the main goal.
+The names are obtained from the respectively named arguments to the tactic -/
+macro "init_next_step" h_run:ident stepi_eq:ident sn:ident : tactic =>
   `(tactic|
-    (rw [run_onestep _ _ _ (by omega)] at $h_run:ident
-    -- I prefer using let instead of obtain because obtain names
-    -- the unsolved goal `case intro`. Then we get `.intro` suffixes
-    -- there every time we run this tactic.
-     let ⟨$sn:ident, ⟨$h_step:ident, _⟩⟩ := $h_run:ident;
-     -- obtain ⟨$sn:ident, ⟨$h_step:ident, $h_run:ident⟩⟩ := $h_run:ident
-     clear $h_run:ident; rename_i $h_run:ident
-     simp (config := {ground := true}) only at $h_run:ident))
+    (-- use `let` over `obtain` to prevent `.intro` goal tags
+     let ⟨$sn:ident, ⟨$stepi_eq:ident, h_run_new⟩⟩ := run_onestep $h_run:ident
+     replace $h_run:ident := h_run_new
+     clear h_run_new
+    ))
 
 section stepiTac
 
-/-- Apply the relevant pre-generated stepi lemma to replace a local hypothesis
-  `h_step : ?s' = stepi ?s`
-with an hypothesis in terms of `w` and `write_mem`
+/-- Apply the relevant pre-generated stepi lemma to a local hypothesis
+  `stepi_eq : stepi ?s = ?s'`
+to obtain a new local hypothesis in terms of `w` and `write_mem`
   `h_step : ?s' = w _ _ (w _ _ (... ?s))`
 -/
-def stepiTac (h_step : Ident) (ctx : SymContext)
+def stepiTac (stepi_eq h_step : Ident) (ctx : SymContext)
   : TacticM Unit := withMainContext do
   let pc := (Nat.toDigits 16 ctx.pc.toNat).asString
   --  ^^ The PC in hex
   let step_lemma := mkIdent <| Name.str ctx.program s!"stepi_eq_0x{pc}"
 
   evalTacticAndTrace <|← `(tactic| (
-    replace $h_step :=
-      _root_.Eq.trans $h_step
+    have $h_step :=
+      _root_.Eq.trans (Eq.symm $stepi_eq)
```

Comment:
Why do we have `stepi ?s = ?s'` instead of the other way round? Does it stem from the new definition of `run_onestep`?